File: Medium/2025-08-06-3478-FruitsIntoBasketsIII/solution.py
from typing import List

class Solution:
    def numOfUnplacedFruits(self, fruits: List[int], baskets: List[int]) -> int:
        min_unplaced = float('inf')
        
        # Try placing all fruits without skipping
        min_unplaced = min(min_unplaced, self.place_fruits(fruits, baskets, -1))
        
        # Try skipping each fruit type
        for skip_index in range(len(fruits)):
            min_unplaced = min(min_unplaced, self.place_fruits(fruits, baskets, skip_index))
        
        return min_unplaced
    
    def place_fruits(self, fruits: List[int], baskets: List[int], skip_index: int) -> int:
        used = [False] * len(baskets)
        unplaced = 0
        
        for i in range(len(fruits)):
            if i == skip_index:
                continue  # Skip this fruit
            
            placed = False
            
            # Find the leftmost available basket with sufficient capacity
            for j in range(len(baskets)):
                if not used[j] and baskets[j] >= fruits[i]:
                    used[j] = True  # Mark basket as used
                    placed = True
                    break
            
            if not placed:
                unplaced += 1
        
        return unplaced 

File: Medium/2025-08-06-3478-FruitsIntoBasketsIII/solution.java
import java.util.*;

class Solution {
    public int numOfUnplacedFruits(int[] fruits, int[] baskets) {
        int minUnplaced = Integer.MAX_VALUE;
        
        // Try placing all fruits without skipping
        minUnplaced = Math.min(minUnplaced, placeFruitsOptimized(fruits, baskets, -1));
        
        // Try skipping each fruit type
        for (int skipIndex = 0; skipIndex < fruits.length; skipIndex++) {
            minUnplaced = Math.min(minUnplaced, placeFruitsOptimized(fruits, baskets, skipIndex));
        }
        
        return minUnplaced;
    }
    
    private int placeFruitsOptimized(int[] fruits, int[] baskets, int skipIndex) {
        // Use TreeSet to maintain available baskets efficiently
        TreeSet<Integer> availableBaskets = new TreeSet<>();
        for (int basket : baskets) {
            availableBaskets.add(basket);
        }
        
        int unplaced = 0;
        
        for (int i = 0; i < fruits.length; i++) {
            if (i == skipIndex) {
                unplaced++; // Count skipped fruit as unplaced
                continue;
            }
            
            // Find the smallest available basket that can hold this fruit
            Integer basket = availableBaskets.ceiling(fruits[i]);
            
            if (basket == null) {
                unplaced++;
            } else {
                availableBaskets.remove(basket);
            }
        }
        
        return unplaced;
    }
} 